rust/sdk/examples/chat/display.rs

```diff
@@ -127,6 +104,200 @@ pub async fn display(mut cell: ActoCell<Display, impl ActoRuntime>) {
+
+        let scroll_range = text_height.checked_sub(message_text_rect_height).unwrap_or(0);
+        let user_scroll_y_from_bottom: u16 = {
+            // Is there a height shift from previous render?
+            let height_shift_on_update = if let Some(prev_height) = scroll_state.last_recorded_par_height {
+                text_height.checked_sub(prev_height).unwrap_or(0)
+            } else {
+                0
+            };
+
+            scroll_state.last_recorded_par_height = Some(text_height);
+
+            // if vscroll_from_bottom is not 0, it must shift up, because UX
+            if scroll_state.vscroll_from_bottom > 0 {
+                scroll_state.vscroll_from_bottom += height_shift_on_update;
+            }
+
+            scroll_state.vscroll_from_bottom = scroll_state.vscroll_from_bottom.clamp(0, scroll_range);
+
+            scroll_state.vscroll_from_bottom
+        };
+
+        let scroll_y = scroll_range.checked_sub(user_scroll_y_from_bottom).unwrap_or(0);
+
+        if text_height > message_text_rect_height {
+            if user_scroll_y_from_bottom != 0 {
+                let visible_bottom = scroll_y + message_text_rect_height - 1;
+                if let Some(last_line) = lines.get_mut(visible_bottom as usize) {
+                    *last_line = Line::from(vec![Span::from("Press Down to scroll down (Hold Ctrl to Boost)")])
+                        .alignment(Alignment::Center);
+                }
+            }
+
+            if scroll_y != 0 {
+                let visible_top = scroll_y;
+                if let Some(first_line) = lines.get_mut(visible_top as usize) {
+                    *first_line = Line::from(vec![Span::from("Press Up to scroll up (Hold Ctrl to Boost)")])
+                        .alignment(Alignment::Center);
+                }
+            }
+        }
+
+        let paragraph = Paragraph::new(lines).dark_gray().scroll((scroll_y, 0));
+
+        f.render_widget(paragraph, message_text_rect);
+    });
+
+    text_area.project(|t| f.render_widget(t.widget(), layout[1]));
+
+    if let Some(ref e) = not_connected {
+        let rect = centered_rect(60, 20, size);
+        let text = Paragraph::new(vec![
+            Line::from(Span::styled(
+                "Not connected to Actyx",
+                Style::new().add_modifier(Modifier::BOLD),
+            )),
+            Line::from(""),
+            Line::from(e.as_str()),
+        ])
+        .alignment(Alignment::Center)
+        .wrap(Wrap { trim: true })
+        .block(
+            Block::default()
+                .title("Error")
+                .borders(Borders::ALL)
+                .style(Style::new().bg(Color::Red)),
+        );
+        f.render_widget(Clear, rect);
+        f.render_widget(text, rect);
+    }
+}
+
+// FIXME sometimes, Scrollbar from ratatui doesn't work for bottom-to-top scrolling
+//  because there is no way to calculate the height of a text
+//  meanwhile Paragraph::wrap has an opaque wrapping mechanism
```

Comment:
how about using the `textwrap` crate:
```rust
fn message_to_lines(msg: &Message, width: u16) -> Vec<Line<'static>> {
    let time = DateTime::<Utc>::from(msg.time).to_rfc3339();
    let mut first = vec![
        Span::styled(time, Style::new().add_modifier(Modifier::ITALIC)),
        Span::raw(" "),
        Span::styled(msg.from.to_owned(), Style::new().add_modifier(Modifier::BOLD)),
        Span::raw(": "),
    ];
    let first_width: usize = first.iter().map(|s| s.width()).sum();
    let indent = Span::raw(" ".repeat(first_width));

    let inputs = textwrap::wrap(&msg.text, usize::from(width) - first_width);
    let mut lines = vec![];
    for (idx, input) in inputs.into_iter().enumerate() {
        let mut v = if idx == 0 {
            std::mem::take(&mut first)
        } else {
            vec![indent.clone()]
        };
        v.push(Span::raw(input.into_owned()));
        lines.push(Line::from(v));
    }
    lines
}
```

Reply:
In the rare case of a very long username, a very narrow terminal, or a very deep zoom, won't that break the bold modifier of the username if it is broken in the middle of the username?

Reply:
made more readable in https://github.com/Actyx/Actyx/pull/554/commits/a52d3233b2e80053dbe04d872393d0b23e2aa816

Reply:
The main point of using `textwrap` is that it handles the fact that not every byte corresponds to a glyph on the screen, and it breaks nicely at whitespace. The other point is to indent all lines after the first. With the current setup (i.e. scroll_x is always zero) a very narrow terminal only leads to not seeing the message at all, which I think is fine. The current format of the timestamp is extremely technical and should probably be shortened to a more usual appearance, but for a demo it is good enough.